### PR TITLE
[REEF-1206]  Adding API to allow deserialize data from remote files directly

### PR DESCRIPTION
**Pull request by jwang98052**
Adding a new API and deprecated the old one for Deserializer
Adding a named parameter for CopyToLocal
Set driver momery for IMRU

JIRA: [REEF-1206](https://issues.apache.org/jira/browse/REEF-1206)

This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs [hunk 1]
```diff
@@ -146,6 +146,7 @@ internal sealed class REEFIMRUClient : IIMRUClient
                 .AddDriverConfiguration(imruDriverConfiguration)
                 .AddGlobalAssemblyForType(typeof(IMRUDriver<TMapInput, TMapOutput, TResult>))
                 .SetJobIdentifier(jobDefinition.JobName)
+                .SetDriverMemory(5000)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs]: This is orthogonal to the change, right?

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs [hunk 2]
```diff
@@ -45,6 +46,7 @@ internal static bool TestWithByteDeserializer()
             var serializerConf = TangFactory.GetTang().NewConfigurationBuilder()
                 .BindImplementation<IFileDeSerializer<IEnumerable<byte>>, ByteSerializer>(GenericType<IFileDeSerializer<IEnumerable<byte>>>.Class,
                     GenericType<ByteSerializer>.Class)
+                .BindNamedParam<CopyToLocal, bool>("true")
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: Is there a way to not use strings for this?

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs [hunk 3]
```diff
@@ -161,5 +163,26 @@ public IEnumerable<byte> Deserialize(string fileFolder)
...
             }
         }
+
+        /// <summary>
+        /// Read bytes from all the files in the set and return one by one
+        /// </summary>
+        /// <param name="filePaths"></param>
+        /// <param name="local"></param>
+        /// <returns></returns>
+        public IEnumerable<byte> Deserialize(ISet<string> filePaths, bool local)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: How is `IEnumerable<byte>` different from `Stream`?

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs [hunk 4]
```diff
@@ -161,5 +163,26 @@ public IEnumerable<byte> Deserialize(string fileFolder)
...
+        {
+            foreach (var f in filePaths)
+            {
+                using (FileStream stream = File.Open(f, FileMode.Open))
+                {
+                    BinaryReader reader = new BinaryReader(stream);
+                    while (reader.PeekChar() != -1)
+                    {
+                        yield return reader.ReadByte();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: this forces us to read a byte at a time? Isn't this very slow?

File: lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs [hunk 5]
```diff
@@ -328,6 +330,27 @@ public IEnumerable<byte> Deserialize(string fileFolder)
...
             }
         }
+
+        /// <summary>
+        /// Enumerate all the files in the set and return each byte read
+        /// </summary>
+        /// <param name="filePaths"></param>
+        /// <param name="local"></param>
+        /// <returns></returns>
+        public IEnumerable<byte> Deserialize(ISet<string> filePaths, bool local)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs]: This is the exact same code as above. Can it be refactored?

File: lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs [hunk 6]
```diff
@@ -371,5 +394,26 @@ public IEnumerable<Row> Deserialize(string fileFolder)
...
             }
         }
+
+        /// <summary>
+        /// read all the files in the set and return byte read one by one
+        /// </summary>
+        /// <param name="filePaths"></param>
+        /// <param name="local"></param>
+        /// <returns></returns>
+        public IEnumerable<Row> Deserialize(ISet<string> filePaths, bool local)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs]: This is the exact same code as above, but it differs in the inner call. Why not return a `Stream` and pass it to a deserializer?

File: lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs [hunk 7]
```diff
@@ -82,17 +85,22 @@ private void Initialize()
         /// <returns></returns>
         public T GetPartitionHandle()
         {
-            if (!_isInitialized)
+            if (_copyToLocal)
             {
-                Initialize();
+                if (!_isInitialized)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs]: Is this thread safe?

File: lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/IFileDeSerializer.cs [hunk 8]
```diff
@@ -30,6 +34,17 @@ public interface IFileDeSerializer<T>
...
+
+        /// <summary>
+        /// The input is a set of file paths. 
+        /// The output is of type T which is defined by the client.
+        /// If there is any IO error, IOException could be thrown.
+        /// </summary>
+        /// <param name="filePaths"></param>
+        /// <param name="local"></param>
+        /// <returns></returns>
+        T Deserialize(ISet<string> filePaths, bool local);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/IFileDeSerializer.cs]: Shouldn't this boolean be passed in as a parameter to the constructor? Is there ever a need to use one and the same instance of `IFileDeSerializer` where it sometimes downloads first and sometimes doesn't?

**Comment by markusweimer** [comment]
I've done a pass.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs]: Right, it is required by IMRU client project. Just don't want to have a PR for this line only. And the change in PR and this line are required by IMRU client project.

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: In the current Tang API, the value of a named parameter for preliminary type like int, bool, are all use string format.

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: We want to return IEnumarable<T>. T can by Row, or int or string.... In this test, I just use byte as an example.

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: It is a test. In the test, I only put a few bytes in total. 
A real case would either an entire file or a big chuck.

File: lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs]: Done. 
The old one will be deprecated next time.

File: lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestFilePartitionInputDataSet.cs]: The previous one will be deprecated.

File: lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemInputPartition.cs]: Yes, lock is used here.

File: lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/IFileDeSerializer.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/IFileDeSerializer.cs]: It is a named parameter from FileSystemInputPartition, then passed to the deserializer. 

Most cases, deserializer only needs one way to get files, wither from remote or copied local files. But if we have a flag, we can easily switch between the two in case we need. We can even  compare the performance between the two.

**Comment by jwang98052** [comment]
@markusweimer Thanks for the review. I have addressed the review comments.

File: lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.TestClient/HadoopFileInputPartitionTest.cs]: ah, fair enough. We probably should fix that in Tang at some point.

**Comment by jwang98052** [comment]
Any more comments @markusweimer ?

**Comment by jwang98052** [comment]
@markusweimer, if you don't have any more comments, I would like to let @tcNickolas to merge for me before the cut off of 0.14 in this weekend.

**Comment by markusweimer** [comment]
I'll test and merge. We can revisit the arch discussion over on JIRA.
